core/LSU.cpp

```diff
@@ -111,13 +111,45 @@ namespace olympia
     // Receive new load/store instruction from Dispatch Unit
     void LSU::getInstsFromDispatch_(const InstPtr & inst_ptr)
     {
-        core_types::RegFile reg_file = core_types::RF_INTEGER;
-        const auto & srcs = inst_ptr->getSourceOpInfoList();
-        if(srcs.size() > 0){
-            reg_file = olympia::coreutils::determineRegisterFile(srcs[0]);
+        bool operands_ready = false;
+        core_types::RegFile reg_file = inst_ptr->getRenameData().getSourceList()[0].rf;
+        if(reg_file == core_types::RegFile::RF_FLOAT){
+            // if we have a floating point store, address is an integer
+            // but data is a float so we need to check both scoreboards
+            uint32_t reg_file_count = 0;
+            for(uint32_t rf = 0; rf < core_types::RegFile::N_REGFILES; ++rf)
+            {
+                reg_file = static_cast<olympia::core_types::RegFile>(rf);
+                const auto & src_bits = inst_ptr->getSrcRegisterBitMask(reg_file);
+                if(scoreboard_views_[reg_file]->isSet(src_bits)){
+                    reg_file_count++;
+                }
+                else{
+                    scoreboard_views_[reg_file]->registerReadyCallback(src_bits, inst_ptr->getUniqueID(),
+                                        [this, inst_ptr](const sparta::Scoreboard::RegisterBitMask&)
+                                        {this->getInstsFromDispatch_(inst_ptr);});
+                    ILOG("Registering Callback: " << inst_ptr);
+                }
+            }
+            if(reg_file_count == core_types::RegFile::N_REGFILES){
+                operands_ready = true;
+            }
+        }
+        else{
+            const auto & src_bits = inst_ptr->getSrcRegisterBitMask(reg_file);
+            if(scoreboard_views_[reg_file]->isSet(src_bits)){
+                operands_ready = true;
+            }
+            else{
+                scoreboard_views_[reg_file]->registerReadyCallback(src_bits, inst_ptr->getUniqueID(),
+                                    [this, inst_ptr](const sparta::Scoreboard::RegisterBitMask&)
+                                    {this->getInstsFromDispatch_(inst_ptr);});
+                ILOG("Registering Callback: " << inst_ptr);
+            }
```

Comment:
Perhaps you see a logic path that I missed, but how about this?  Yeah, it's a bit redundant (will check INT twice for example), but it's a good start...

```
bool all_ready = true; // assume all ready
for(const auto & src_reg : inst_ptr->getRenameData().getSourceList())
{
    const auto & src_bits = inst_ptr->getSrcRegisterBitMask(src_reg.rf);
    if(false == scoreboard_views_[src_reg.rf]->isSet(src_bits)) {
        all_ready = false;
        break;
    }
}

if (all_ready) {
    // Create load/store memory access info
...
```

Reply:
Yeah this looks a lot cleaner compared to what I have, will use your implementation instead and test.

Reply:
Implemented in [4189f59](https://github.com/riscv-software-src/riscv-perf-model/pull/70/commits/4189f590b3157d70f90ed749a664447fd88080ea)